text/0000-array-private-classes

```diff
@@ -0,0 +1,148 @@
+- Feature Name: collections should not expose internal classes through iterators
+- Start Date: 2022-01-08
+- RFC PR: (leave this empty)
+- Pony Issue: (leave this empty)
+
+# Summary
+
+Following information hiding design principle, the builtin classes of collection data structures must not be made visible through iterator functions like `ArrayKeys`, `ArrayValues` and `ArrayPairs`. These classes can be made private as they are only used as return types for `Array` functions `keys`, `values` and `pairs`. The return values for these functions are changed to the more general interface `Iterator`.
+
+A new interface `RewindableIterator` is defined to allow for rewindable iterators, like it is the case for `Array` `values`.
+
+This design principle is applied to the other collection classes that expose internals too like:
+
+* `List`
+* `Map`
+* persistent `Map`
+* `Vec`
+* persistent `Vec`
+* `Set`
+* `Itertools`
+
+# Motivation
+
+This change brings:
+
+- Applying the design principle of [hiding implementation details](https://en.wikipedia.org/wiki/Information_hiding) but offer a general and stable interface. Returning interfaces instead of concrete classes allows changing the implementation. Usually, one must return the most general type that fullfils the contract of the function (in the case of the functions discussed in this RFC, iteration).
+- Collections' functions `keys`, `values` and `pairs` definitions are made more general. Iterators implementation details are not public. Internal classes used by implementation like `*Keys`, `*Values` and `*Pairs` are now [opaque data types](https://en.wikipedia.org/wiki/Opaque_data_type). Generally, when using these collection classes, clients are not interested by the iterators implementation, but by the types these iterators return and that is provided by the generic parameters.
+- The generic return signature of these 3 iterating functions is simpler to understand for clients of collection classes.
+- Makes the standard library more simple by hiding 18 specialised classes from stdlib of which 3 are from `builtin`.
+- The interface `RewindableIterator` is added to create rewindable iterators (can be re-start from first value).
+
+This change remains compatible with the existing code base but for client code that is directly using the classes `*Keys`, `*Values` and `*Pairs`. A search on Github shows that the impact is very limited.
```

Comment:
There seems to be a thought missing from:

`
This change remains compatible with the existing code base but for client code that is directly using the classes `*Keys`, `*Values` and `*Pairs`.
`

What did you mean to say here? That this is a breaking change for anyone who names any of the concrete implementations in their code?

If yes, that isn't really a motivation. That would be a drawback. I don't believe you intend to say that breaking people's existing code is a motivation for this change.

Reply:
I don't know how to translate **but for client** when you want to say that only that minority of users will be impacted by this change...

Show me actual code that will break.

Reply:
I think this there's nothing here that is a motivation and you should either cut the text in question or move it summary.

Having a summary that gives a quick overview and says that it is a breaking change but you expect from a GitHub search that it should impact only a small amount of code would be reasonable.

Reply:
Done